File: backend/metateam/runtime/think_tags.py
"""Split <think>…</think> / <thinking>…</thinking> from model content streams."""

from __future__ import annotations

import re
from typing import Iterable

_THINK_OPEN = re.compile(r"<(think|thinking)\s*>", re.IGNORECASE)
_THINK_CLOSE = re.compile(r"</(think|thinking)\s*>", re.IGNORECASE)
_OPEN_PREFIXES = (
    "<",
    "<t",
    "<th",
    "<thi",
    "<thin",
    "<think",
    "<thinking",
    "<think>",
    "<thinking>",
)
_CLOSE_PREFIXES = (
    "<",
    "</",
    "</t",
    "</th",
    "</thi",
    "</thin",
    "</think",
    "</thinking",
    "</think>",
    "</thinking>",
)


def _partial_suffix_len(buf: str, prefixes: tuple[str, ...]) -> int:
    if not buf:
        return 0
    low = buf.lower()
    max_keep = 0
    for i in range(1, min(len(low), 12) + 1):
        suffix = low[-i:]
        if any(p.startswith(suffix) for p in prefixes):
            max_keep = i
    return max_keep


def split_think_tags(text: str) -> tuple[str, str]:
    """Split a complete string into (visible_content, thinking_content)."""
    if not text:
        return "", ""
    splitter = ThinkTagSplitter()
    content_parts: list[str] = []
    reason_parts: list[str] = []
    for kind, piece in list(splitter.feed(text)) + list(splitter.flush()):
        if kind == "reasoning":
            reason_parts.append(piece)
        else:
            content_parts.append(piece)
    return "".join(content_parts), "".join(reason_parts)


class ThinkTagSplitter:
    """Incremental splitter for streamed content that embeds think tags.

    Tag interiors are treated as reasoning; everything outside is visible content.
    """

    def __init__(self) -> None:
        self._buf = ""
        self._in_think = False

    def feed(self, text: str) -> list[tuple[str, str]]:
        if not text:
            return []
        self._buf += text
        out: list[tuple[str, str]] = []
        while True:
            if not self._in_think:
                m = _THINK_OPEN.search(self._buf)
                if not m:
                    keep = _partial_suffix_len(self._buf, _OPEN_PREFIXES)
                    emit = self._buf[:-keep] if keep else self._buf
                    self._buf = self._buf[-keep:] if keep else ""
                    if emit:
                        out.append(("content", emit))
                    break
                before = self._buf[: m.start()]
                if before:
                    out.append(("content", before))
                self._buf = self._buf[m.end() :]
                self._in_think = True
            else:
                m = _THINK_CLOSE.search(self._buf)
                if not m:
                    keep = _partial_suffix_len(self._buf, _CLOSE_PREFIXES)
                    emit = self._buf[:-keep] if keep else self._buf
                    self._buf = self._buf[-keep:] if keep else ""
                    if emit:
                        out.append(("reasoning", emit))
                    break
                inside = self._buf[: m.start()]
                if inside:
                    out.append(("reasoning", inside))
                self._buf = self._buf[m.end() :]
                self._in_think = False
        return out

    def flush(self) -> list[tuple[str, str]]:
        if not self._buf:
            return []
        kind = "reasoning" if self._in_think else "content"
        piece = self._buf
        self._buf = ""
        return [(kind, piece)] if piece else []

    def reset(self) -> None:
        self._buf = ""
        self._in_think = False

File: ui/src/utils/thinkTags.ts
/** Incremental splitter for <think>…</think> / <thinking>…</thinking> in streamed content. */

const OPEN_RE = /<(think|thinking)\s*>/i;
const CLOSE_RE = /<\/(think|thinking)\s*>/i;
const OPEN_PREFIXES = [
  "<",
  "<t",
  "<th",
  "<thi",
  "<thin",
  "<think",
  "<thinking",
  "<think>",
  "<thinking>",
];
const CLOSE_PREFIXES = [
  "<",
  "</",
  "</t",
  "</th",
  "</thi",
  "</thin",
  "</think",
  "</thinking",
  "</think>",
  "</thinking>",
];

function partialSuffixLen(buf: string, prefixes: string[]): number {
  if (!buf) return 0;
  const low = buf.toLowerCase();
  let maxKeep = 0;
  const n = Math.min(low.length, 12);
  for (let i = 1; i <= n; i++) {
    const suffix = low.slice(-i);
    if (prefixes.some((p) => p.startsWith(suffix))) maxKeep = i;
  }
  return maxKeep;
}

export type ThinkPiece = { kind: "content" | "reasoning"; text: string };

export class ThinkTagSplitter {
  private buf = "";
  private inThink = false;

  feed(text: string): ThinkPiece[] {
    if (!text) return [];
    this.buf += text;
    const out: ThinkPiece[] = [];
    while (true) {
      if (!this.inThink) {
        const m = OPEN_RE.exec(this.buf);
        if (!m) {
          const keep = partialSuffixLen(this.buf, OPEN_PREFIXES);
          const emit = keep ? this.buf.slice(0, -keep) : this.buf;
          this.buf = keep ? this.buf.slice(-keep) : "";
          if (emit) out.push({ kind: "content", text: emit });
          break;
        }
        const before = this.buf.slice(0, m.index);
        if (before) out.push({ kind: "content", text: before });
        this.buf = this.buf.slice(m.index + m[0].length);
        this.inThink = true;
      } else {
        const m = CLOSE_RE.exec(this.buf);
        if (!m) {
          const keep = partialSuffixLen(this.buf, CLOSE_PREFIXES);
          const emit = keep ? this.buf.slice(0, -keep) : this.buf;
          this.buf = keep ? this.buf.slice(-keep) : "";
          if (emit) out.push({ kind: "reasoning", text: emit });
          break;
        }
        const inside = this.buf.slice(0, m.index);
        if (inside) out.push({ kind: "reasoning", text: inside });
        this.buf = this.buf.slice(m.index + m[0].length);
        this.inThink = false;
      }
    }
    return out;
  }

  flush(): ThinkPiece[] {
    if (!this.buf) return [];
    const kind = this.inThink ? "reasoning" : "content";
    const text = this.buf;
    this.buf = "";
    return text ? [{ kind, text }] : [];
  }

  reset() {
    this.buf = "";
    this.inThink = false;
  }
}

/** Non-streaming split of a complete string. */
export function splitThinkTags(text: string): { content: string; reasoning: string } {
  const s = new ThinkTagSplitter();
  const content: string[] = [];
  const reasoning: string[] = [];
  for (const p of [...s.feed(text), ...s.flush()]) {
    if (p.kind === "reasoning") reasoning.push(p.text);
    else content.push(p.text);
  }
  return { content: content.join(""), reasoning: reasoning.join("") };
}
